src/GNDStk/JSON/test/JSON.test.cpp

```diff
@@ -4,6 +4,317 @@
+   WHEN("We construct a JSON from a file") {
+      const JSON j("n-069_Tm_170-covar.json");
+      THEN("It should produce an equivalent to the tree made from the file") {
+         CHECK(!j.empty());
+         std::ostringstream oss1; oss1 << tree("n-069_Tm_170-covar.json");
+         std::ostringstream oss2; oss2 << tree(j);
+         CHECK(oss1.str() == oss2.str());
+      }
+   }
+
+   // from istream
+   WHEN("We construct a JSON from an istream") {
+      std::ifstream ifs("n-069_Tm_170-covar.json");
+      const JSON j(ifs);
+      THEN("It should produce an equivalent to the tree made from the file") {
+         CHECK(!j.empty());
+         std::ostringstream oss1; oss1 << tree("n-069_Tm_170-covar.json");
+         std::ostringstream oss2; oss2 << tree(j);
+         CHECK(oss1.str() == oss2.str());
+      }
+   }
+
+
+   // ------------------------
+   // assignment
+   // ------------------------
+
+   // move
+   WHEN("We call JSON's move assignment") {
+      JSON j;
+      j = JSON{"n-069_Tm_170-covar.json"};
+      CHECK(!j.empty());
+      // and ensure it moved correctly
+      std::ostringstream oss1; oss1 << JSON{"n-069_Tm_170-covar.json"};
+      std::ostringstream oss2; oss2 << j;
+      CHECK(oss1.str() == oss2.str());
+   }
+
+   // copy
+   WHEN("We call JSON's copy assignment") {
+      const JSON a("n-069_Tm_170-covar.json");
+      JSON j;
+      j = a;
+      CHECK(!j.empty());
+      // and ensure it copied correctly
+      std::ostringstream oss1; oss1 << a;
+      std::ostringstream oss2; oss2 << j;
+      CHECK(oss1.str() == oss2.str());
+   }
+
+
+   // ------------------------
+   // read
+   // ------------------------
+
+   // from istream
+   WHEN("We read a JSON from an istream") {
+      std::ifstream ifs("n-069_Tm_170-covar.json");
+      const JSON j(ifs);
+      THEN("It should produce an equivalent to the tree made from the file") {
+         CHECK(!j.empty());
+         std::ostringstream oss1; oss1 << tree("n-069_Tm_170-covar.json");
+         std::ostringstream oss2; oss2 << tree(j);
+         CHECK(oss1.str() == oss2.str());
+      }
+   }
+
+   // from file
+   WHEN("We read a JSON from a file") {
+      const JSON j("n-069_Tm_170-covar.json");
+      THEN("It should produce an equivalent to the tree made from the file") {
+         CHECK(!j.empty());
+         std::ostringstream oss1; oss1 << tree("n-069_Tm_170-covar.json");
+         std::ostringstream oss2; oss2 << tree(j);
+         CHECK(oss1.str() == oss2.str());
+      }
+   }
+
+
+   // ------------------------
+   // write
+   // ------------------------
+
+   // to ostream
+   WHEN("We write a JSON to an ostream") {
+      const JSON j("n-069_Tm_170-covar.json");
+      std::ostringstream oss;
+      j.write(oss);
+      THEN("We can check that the result is what we expect") {
+         CHECK(oss.str() == string_real_json);
+      }
+   }
+
+   // to file
+   // fixme How do we handle *writing* a file with this test system?
```

Comment:
We write functionality to write to a `std::ostream` or `std::string` object and test that functionality. Writing to a file after you have a string is trivial. ENDFtk doesn't ever write to a file. You just give it an iterator to a string and it starts writing there.

Reply:
It's certainly convenient for users, however, to provide an API that allows direct file writing. I provide functionality to do that, as well as to write to `ostream`. Then, the file writing capability actually just uses an `ofstream` and calls the `ostream` version. In my opinion, one would still ideally test the writing capabilities to write directly to a file, but if it's fine not to (I know it works), that that's good as well. I can just delete the "fixme" you referenced. Deleting fixmes, without having to do any work, isn't something I'll object to. :-)